include/mbedtls/pk.h

```diff
@@ -484,6 +484,109 @@ int mbedtls_pk_can_do_ext(const mbedtls_pk_context *ctx, psa_algorithm_t alg,
                           psa_key_usage_t usage);
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
 
+#if defined(MBEDTLS_PSA_CRYPTO_CLIENT)
+/**
+ * \brief           Determine valid PSA attributes that can be used to
+ *                  import a key into PSA.
+ *
+ *                  The attributes determined by this function are suitable
+ *                  for calling mbedtls_pk_import_into_psa() to create
+ *                  a PSA key with the same key material.
+ *
+ *                  The typical flow of operations involving this function is
+ *                  ```
+ *                  psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
+ *                  int ret = mbedtls_pk_guess_psa_attributes(pk, &attributes);
+ *                  if (ret != 0) ...; // error handling omitted
+ *                  // Tweak attributes if desired
+ *                  psa_key_id_t key_id = 0;
+ *                  ret = mbedtls_pk_import_into_psa(pk, &attributes, &key_id);
+ *                  if (ret != 0) ...; // error handling omitted
+ *                  ```
+ *
+ * \note            This function does not support RSA-alt contexts
+ *                  (set up with mbedtls_pk_setup_rsa_alt()).
+ *
+ * \param[in] pk    The PK context to use. It must have been set up.
+ *                  It can either contain a key pair or just a public key.
+ * \param[out] attributes
+ *                  On success, valid attributes to import the key into PSA.
+ *                  - The lifetime and key identifier are unchanged. If the
+ *                    attribute structure was initialized or reset before
+ *                    calling this function, this will result in a volatile
+ *                    key. Call psa_set_key_identifier() after this
+ *                    function if you wish to create a persistent key. Call
+ *                    psa_set_key_lifetime() after this function if
+ *                    you wish to import the key in a secure element.
+ *                  - The key type and bit-size are determined by the contents
+ *                    of the PK context. The key type is a key pair type
+ *                    (#PSA_KEY_TYPE_RSA_KEY_PAIR or
+ *                    #PSA_KEY_TYPE_ECC_KEY_PAIR(...)) if the PK context
+ *                    contains a private key, and a public key type
+ *                    (#PSA_KEY_TYPE_RSA_PUBLIC_KEY or
+ *                    #PSA_KEY_TYPE_ECC_PUBLIC_KEY(...)) otherwise.
+ *                  - The permitted algorithm is as follows (for signature
```

Comment:
I like the idea of an extra argument indicating encryption/signature/agreement. I think it would a good balance, requesting from the user just the information we need to then set appropriate policies.

Reply:
In writing up the extra argument, I started thinking that maybe that argument could also indicate “I just want the public key” by passing VERIFY/ENCRYPT, or “I want the key pair” by passing SIGN/DECRYPT. And that makes it a subset of `psa_key_usage_flags_t`, saving us from inventing a new enum.

I'm not completely convinced by this train of thought, in part because there's no way to say “I only want the public key” for a `PK_ECKEY_DH`. But that's not a common use case, so I'm not sure if this wart is really important.